Avoid SIGSEGV on shutdown with target re-resolution

michaelweiser opened the pull request:
spiped will segfault on shutdown if target address re-resolution is in
use. This is due to two cases of accessing memory that will be freed by
other parts of the code on the first re-resolution cycle. To avoid this:

The DNS timer has to be marked as expired so it's not freed on shutdown
if already expired. Incidentally, the new instance's is recorded so it
can be cancelled on shutdown if not already expired.

Also, the initial target address list as taken from the command line has
to be marked as under dispatch's purview after dispatch_accept() has
been called.

michaelweiser commented:
Travis build failure on macOS is unrelated. Fix here: https://github.com/Tarsnap/spiped/pull/154.

gperciva commented:
Thanks, @michaelweiser.  Could you please rebase onto current master and do a force push to trigger a new Travis-CI build?

gperciva commented on spiped/dispatch.c:
```diff
@@ -61,8 +61,9 @@ callback_resolve(void * cookie, struct sock_addr ** sas)
-	    A, A->rtime) == NULL)
+	A->dnstimer_cookie = events_timer_register_double(callback_resolveagain,
+	    A, A->rtime);
+	if (A->dnstimer_cookie == NULL)
```
Our code style would write this as:
```
if ((A->dnstimer_cookie = events_blah(blah)) == NULL)
```
I personally prefer your version, but I know that it won't be accepted as-is.  Also, having a consistent style is more important than any one particular stylistic preference.  :)

gperciva commented on spiped/dispatch.c:
```diff
@@ -79,6 +80,10 @@ callback_resolveagain(void * cookie)
 {
 	struct accept_state * A = cookie;
 
+	/* this timer is expired for sure  - mark it as such so we don't try to
```
For a multiline comment, please leave the `/*` and `*/` on separate lines:
```
/*
 * This timer is expired - mark it as such so
 * we don't try to cancel it on shutdown.
 */
```

gperciva commented:
Hi @michaelweiser!  I made some sample changes (the style of comments), and uploaded to https://github.com/Tarsnap/spiped/tree/michaelweiser-master.  If you like them, please rebase onto your patch (and the whole thing on to current master).

michaelweiser commented:
Hi @gperciva, all done as you suggested. I squashed your style changes into my commit for atomicity and rebased onto current master. Travis is happy now, too.

gperciva commented:
Thanks, LGTM!  I know that @cperciva is busy preparing a paper for a conference, so no guarantees as to when he can take a look at it, but I'll make sure it's in the first batch.

michaelweiser commented:
No prob whatsover: Over debugging I realized I don't even need the re-resolution feature since my spiped only ever connects to localhost and just disabled it. Just wanted to share my findings so the next poor soul doesn't run into it. :)

gperciva commented on spiped/main.c:
```diff
@@ -341,7 +344,8 @@ main(int argc, char * argv[])
 
 	/* Free arrays of resolved addresses. */
-	sock_addr_freelist(sas_t);
+	if (sas_t != NULL)
```
I just noticed that `sock_addr_freelist()` does nothing if the pointer is already NULL, so this line isn't strictly necessary.

gperciva commented on spiped/main.c:
```diff
@@ -357,7 +361,8 @@ main(int argc, char * argv[])
 	free(K);
 err3:
-	sock_addr_freelist(sas_t);
+	if (sas_t != NULL)
```
I just noticed that `sock_addr_freelist()` does nothing if the pointer is already NULL, so this line isn't strictly necessary.

michaelweiser commented:
Right on!

cperciva commented:
Seems like the `sock_addr_freelist(sas_t);` in main.c could just be unconditionally removed now?  Is there any way for `sas_t` to *not* be NULL when we get there?
